integration_tests/framework/xml_test_runner.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Test main and flags."""
+
+import logging
+import os
+import sys
+import time
+import traceback
+import unittest
+
+from xml.etree import ElementTree
+
+XML_OUTPUT_FILE = 'XML_OUTPUT_FILE'
+DEFAULT_SUBTEST = 'Test Case'
+
+
+def _GetMillis():
+  return int(time.time() * 1000)
+
+
+def _Typename(cls):
+  return '%s.%s' % (cls.__module__, cls.__name__)
+
+
+def _AddProperties(element, props):
+  if props:
+    sub = ElementTree.SubElement(element, 'properties')
+    for k, v in props.items():
+      ElementTree.SubElement(sub, 'property', name=k, value=v)
+
+
+class XmlTestRunner(object):
+  """Runner producing test xml output."""
+
+  def __init__(self, flags, filename=None):
+    if not filename:
+      if XML_OUTPUT_FILE in os.environ:
+        filename = os.environ[XML_OUTPUT_FILE]
+    self.filename = filename
+    self.flags = flags
+
+  def run(self, test):  # pylint: disable=invalid-name
+    result = XmlTestResult(self.flags)
+    logging.info('Running tests...')
+    test(result)
+    logging.info('\n\n******************* TESTS FINISHED *******************\n')
+    logging.info('Ran %d tests with %d errors and %d failures', result.testsRun,
+                 len(result.errors), len(result.failures))
+    result.test_suite_element.set('tests', str(result.testsRun))
+    result.test_suite_element.set(
+        'failures', str(len(result.errors) + len(result.failures)))
+    result.test_suite_element.set('time', str(_GetMillis() - result.start_time))
+    if self.filename:
+      with open(self.filename, 'wb') as f:
+        ElementTree.ElementTree(element=result.root_element).write(f)
+    else:
+      for test_and_stack in result.failures:
+        logging.info('\n\n[  TEST FAILED  ] %s\n', test_and_stack[0])
+        logging.info(
+            '%s', test_and_stack[1].replace('\\\\r', '\r').replace(
+                '\\\\n', '\n').replace('\\r', '\r').replace('\\n', '\n'))
+
+    return result
+
+
+class XmlTestResult(unittest.TestResult):
```

Comment:
Same here, would be nice if you describe how the XML file is intended to be used.